# 한국주식 주가 데이터 수집하기 with FinanceDataReader, pykrx, pandas_datareader, yfinance


In [1]:
import FinanceDataReader as fdr
from pykrx import stock
import pandas_datareader.data as pdr
import yfinance as yf

# 1. 종목 리스트 확인하기

+ FinanceDataReader와 pykrx는 종목리스트를 확인할 수 있는 함수 존재

## 1.1. FinanceDataReader

+ 현재 시점의 시장별 상장종목 리스트를 가져올 수 있음
+ 종목코드, 시장, 종목명, 섹터, 산업군, 상장일, 결산월, 대표자, 홈페이지, 사업체 지역을 알 수 있음
+ 과거 시점의 상장종목 리스트는 알 수 없으나 상장폐지 종목을 조회할 수 있는 기능을 별도로 제공

In [2]:
# 가장 최근 영업일의 시장별 종목리스트를 가져옴
stocks = fdr.StockListing('KRX') # 코스피, 코스닥, 코넥스 전체
# stocks = fdr.StockListing('KOSPI') # 코스피
# stocks = fdr.StockListing('KOSDAQ') # 코스닥
# stocks = fdr.StockListing('KONEX') # 코넥스

In [3]:
stocks.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7895 entries, 0 to 7894
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Symbol          7895 non-null   object        
 1   Market          7895 non-null   object        
 2   Name            7895 non-null   object        
 3   Sector          2500 non-null   object        
 4   Industry        2482 non-null   object        
 5   ListingDate     2500 non-null   datetime64[ns]
 6   SettleMonth     2500 non-null   object        
 7   Representative  2500 non-null   object        
 8   HomePage        2349 non-null   object        
 9   Region          2500 non-null   object        
dtypes: datetime64[ns](1), object(9)
memory usage: 678.5+ KB


In [4]:
stocks

Symbol  Market             Name                    Sector  \
0     060310  KOSDAQ               3S                  전자부품 제조업   
1     095570   KOSPI           AJ네트웍스           산업용 기계 및 장비 임대업   
2     006840   KOSPI            AK홀딩스                    기타 금융업   
3     054620  KOSDAQ           APS홀딩스                    기타 금융업   
4     265520  KOSDAQ            AP시스템             특수 목적용 기계 제조업   
...      ...     ...              ...                       ...   
7890  000545   KOSPI            흥국화재우                       NaN   
7891  003280   KOSPI             흥아해운                    해상 운송업   
7892  037440  KOSDAQ               희림  건축기술, 엔지니어링 및 관련 기술 서비스업   
7893  419890   KOSPI  히어로즈 단기채권ESG액티브                       NaN   
7894  238490  KOSDAQ               힘스             특수 목적용 기계 제조업   

                            Industry ListingDate SettleMonth  \
0                        반도체 웨이퍼 캐리어  2002-04-23         03월   
1                렌탈(파렛트, OA장비, 건설장비)  2015-08-21         12월   
2                               지주사업  1999-08-11         12월   
3                        인터넷 트래픽 솔루션  2001-12-04         12월   
4                        디스플레이 제조 장비  2017-04-07         12월   
...                              ...         ...         ...   
7890                             NaN         NaT         NaN   
7891                  외항화물운송업(케미컬탱커)  1976-06-29         12월   
7892                       설계 및 감리용역  2000-02-03         12월   
7893                             NaN         NaT         NaN   
7894  OLED Mask 인장기, OLED Mask 검사기 등  2017-07-20         12월   

            Representative                       HomePage Region  
0                      김세완           http://www.3sref.com  서울특별시  
1                      박대현         http://www.ajnet.co.kr  서울특별시  
2        채형석, 이석주(각자 대표이사)  http://www.aekyunggroup.co.kr  서울특별시  
3                      정기로   http://www.apsholdings.co.kr    경기도  
4                      김영주     http://www.apsystems.co.kr    경기도  
...                    ...                            ...    ...  
7890                   NaN                            NaN    NaN  
7891                   이환구         http://www.heung-a.com  서울특별시  
7892  정영균, 이목운, 허철호 (각자대표)          http://www.heerim.com  서울특별시  
7893                   NaN                            NaN    NaN  
7894                   김주환          http://www.hims.co.kr  인천광역시  

[7895 rows x 10 columns]

In [5]:
type(stocks)

pandas.core.frame.DataFrame

In [64]:
# KRX stock delisting symbol list 상장폐지 종목 전체 리스트
krx_delisting = fdr.StockListing('KRX-DELISTING')
krx_delisting

Symbol         Name  Market SecuGroup Kind ListingDate DelistingDate  \
0     06031012         3S R  KOSDAQ   신주인수권증서  보통주  2012-05-14    2012-05-21   
1     00684014      AK홀딩스8R   KOSPI   신주인수권증서  보통주  2014-07-28    2014-08-04   
2     13893015   BNK금융지주 8R   KOSPI   신주인수권증서  보통주  2015-12-24    2016-01-05   
3     13893014     BS금융지주5R   KOSPI   신주인수권증서  보통주  2014-06-18    2014-06-25   
4     03204017  C&S자산관리 34R  KOSDAQ   신주인수권증서  보통주  2017-02-02    2017-02-09   
...        ...          ...     ...       ...  ...         ...           ...   
3350    034370         럭키소재   KOSPI        주권       1979-06-23    1991-11-11   
3351    028460        태평양건설   KOSPI        주권       1977-06-09    1991-10-05   
3352    028450       금성투자금융   KOSPI        주권       1986-08-29    1991-09-02   
3353    028440           삼화   KOSPI        주권       1974-06-12    1991-07-12   
3354    029260         금성전기   KOSPI        주권       1976-06-29    1991-06-13   

            Reason ArrantEnforceDate ArrantEndDate Industry  ParValue  \
0                                NaT           NaT                0.0   
1                                NaT           NaT                0.0   
2                                NaT           NaT                0.0   
3                                NaT           NaT                0.0   
4                                NaT           NaT                0.0   
...            ...               ...           ...      ...       ...   
3350      해산 사유 발생               NaT           NaT                NaN   
3351  영업활동정지 6월 계속        1991-08-28    1991-10-04                NaN   
3352      해산 사유 발생               NaT           NaT                NaN   
3353     감사의견 의견거절        1991-06-05    1991-07-11                NaN   
3354      해산 사유 발생               NaT           NaT                NaN   

      ListingShares ToSymbol ToName  
0         1194422.0                  
1         1278299.0                  
2        55969410.0                  
3        32791220.0                  
4         3995063.0                  
...             ...      ...    ...  
3350            NaN                  
3351            NaN                  
3352            NaN                  
3353            NaN                  
3354            NaN                  

[3355 rows x 15 columns]

In [131]:
krx_delisting.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3355 entries, 0 to 3354
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Symbol             3355 non-null   object        
 1   Name               3355 non-null   object        
 2   Market             3355 non-null   object        
 3   SecuGroup          3355 non-null   object        
 4   Kind               3355 non-null   object        
 5   ListingDate        3355 non-null   datetime64[ns]
 6   DelistingDate      3355 non-null   datetime64[ns]
 7   Reason             3355 non-null   object        
 8   ArrantEnforceDate  989 non-null    datetime64[ns]
 9   ArrantEndDate      989 non-null    datetime64[ns]
 10  Industry           3355 non-null   object        
 11  ParValue           3124 non-null   float64       
 12  ListingShares      3307 non-null   float64       
 13  ToSymbol           3355 non-null   object        
 14  ToName  

## 1.2. pykrx

+ 특정 시점의 종목코드 조회가능
+ 종목코드에 해당하는 종목명을 조회하는 함수가 별도로 존재함

In [68]:
# 날짜를 명시해주지 않으면 가장 최근 영업일의 시장별 종목리스트를 가져옴
tickers = stock.get_market_ticker_list()
tickers[:6], tickers[-6:]

(['095570', '006840', '027410', '282330', '138930', '001460'],
 ['005010', '069260', '000540', '000547', '000545', '003280'])

In [69]:
tickers = stock.get_market_ticker_list('2010-01-01')
tickers[:6], tickers[-6:]

(['004560', '004565', '001460', '001465', '084680', '001040'],
 ['005010', '069260', '000540', '000547', '000545', '003280'])

In [65]:
stock.get_market_ticker_name('005930')

'삼성전자'

+ 종목리스트를 조회할 때는 FinanceDataReader의 `StockListing` 함수를 이용하는 것이 좋을듯

# 2. 상장종목의 주가 조회

+ 삼성전자(005930)의 주가 조회
+ [삼성전자의 상장일은 1975-06-11](http://asp01.fnguide.com/SVO2/asp/SVD_Corp.asp?pGB=1&gicode=A005930&cID=&MenuYn=Y&ReportGB=&NewMenuID=102&stkGb=752)
+ 각 API를 통해 어느 시점부터 데이터를 조회할 수 있는지 비교

In [6]:
start_date = '1975-06-11'
end_date = '2022-03-23'

## 2.1. FinanceDataReader

+ 시간이 지날수록 조회할 수 있는 최대 과거 시점이 점점 뒤로 밀림
+ 예를 들어, 2022-03-22에 조회했을 때는 1998-02-05 부터 조회됐었는데, 2022-03-23에 조회했을 때는 1998-02-06 부터 조회됨
+ 수정주가로 조회됨

In [7]:
df_fdr = fdr.DataReader('005930', start=start_date, end=end_date)
df_fdr

Open   High    Low  Close    Volume    Change
Date                                                      
1998-02-07   1441   1552   1425   1530    713950       NaN
1998-02-09   1525   1595   1508   1589    965819  0.038562
1998-02-10   1583   1613   1470   1489    549720 -0.062933
1998-02-11   1488   1488   1396   1471    576190 -0.012089
1998-02-12   1452   1523   1416   1507    745520  0.024473
...           ...    ...    ...    ...       ...       ...
2022-03-17  71200  71800  70900  71200  17646315  0.011364
2022-03-18  70600  70900  70200  70700  14410038 -0.007022
2022-03-21  70900  71000  69900  69900  11169002 -0.011315
2022-03-22  69900  70500  69900  70300   9402666  0.005722
2022-03-23  70600  71200  70300  70500  12398025  0.002845

[5999 rows x 6 columns]

## 2.2. pykrx

+ 1990년 데이터부터 조회 가능
+ 수정주가가 디폴트로 조회됨(`adjsted`=True)
+ `adjusted`=False로 조회하면 1995-05-02부터 조회되고, 거래대금과 등락률 칼럼도 붙어있음

In [82]:
df_pykrx = stock.get_market_ohlcv_by_date(fromdate=start_date, 
                                          todate=end_date,
                                          ticker="005930")
df_pykrx

시가     고가     저가     종가       거래량
날짜                                              
1990-01-03  44000  45000  43200  44800     26240
1990-01-04  45000  45800  44800  45300     41900
1990-01-05  45000  45300  44300  44300     27400
1990-01-06  44800  45000  44500  44500     26380
1990-01-08  44500  44900  44000  44000     14790
...           ...    ...    ...    ...       ...
2022-03-17  71200  71800  70900  71200  17646315
2022-03-18  70600  70900  70200  70700  14410038
2022-03-21  70900  71000  69900  69900  11169002
2022-03-22  69900  70500  69900  70300   9402666
2022-03-23  70600  71200  70300  70500  12332242

[8373 rows x 5 columns]

In [86]:
df_pykrx = stock.get_market_ohlcv_by_date(fromdate=start_date, 
                                          todate=end_date,
                                          ticker="005930", adjusted=True)
df_pykrx

시가     고가     저가     종가       거래량
날짜                                              
1990-01-03  44000  45000  43200  44800     26240
1990-01-04  45000  45800  44800  45300     41900
1990-01-05  45000  45300  44300  44300     27400
1990-01-06  44800  45000  44500  44500     26380
1990-01-08  44500  44900  44000  44000     14790
...           ...    ...    ...    ...       ...
2022-03-17  71200  71800  70900  71200  17646315
2022-03-18  70600  70900  70200  70700  14410038
2022-03-21  70900  71000  69900  69900  11169002
2022-03-22  69900  70500  69900  70300   9402666
2022-03-23  70600  71200  70300  70500  12332242

[8373 rows x 5 columns]

In [87]:
df_pykrx = stock.get_market_ohlcv_by_date(fromdate=start_date, 
                                          todate=end_date,
                                          ticker="005930", adjusted=False)
df_pykrx

시가      고가      저가      종가       거래량           거래대금   등락률
날짜                                                                       
1995-05-02  121000  121000  118500  119500    139560    16676735000 -0.83
1995-05-03  119500  126500  119500  123500    382980    47649710000  3.35
1995-05-04  124000  124500  122000  122500    175590    21609210000 -0.81
1995-05-06  122000  123500  122000  122000     47440     5810535000 -0.41
1995-05-08  122000  122000  120500  121000     91810    11127115000 -0.82
...            ...     ...     ...     ...       ...            ...   ...
2022-03-17   71200   71800   70900   71200  17646315  1258698481050  1.14
2022-03-18   70600   70900   70200   70700  14410038  1017183110700 -0.70
2022-03-21   70900   71000   69900   69900  11169002   784175429300 -1.13
2022-03-22   69900   70500   69900   70300   9402666   660385691100  0.57
2022-03-23   70600   71200   70300   70500  12398025   875833161100  0.28

[6808 rows x 7 columns]

## 2.3. pandas_datareader

+ 데이터소스를 naver로 지정하면 1990년 데이터부터, yahoo로 지정하면 2000년 데이터부터 조회됨
+ 수정주가로 조회되는데 adjust 옵션은 따로 없는 듯
+ yahoo finance를 데이터 소스로 사용할 때는 종목코드 뒤에 코스피 종목인 경우 .KS, 코스닥 종목인 경우 .KQ를 붙여줘야 함

In [88]:
df_pdr = pdr.DataReader('005930', 'naver', start=start_date, end=end_date)
df_pdr

Open   High    Low  Close    Volume
Date                                            
1990-01-03  44000  45000  43200  44800     26240
1990-01-04  45000  45800  44800  45300     41900
1990-01-05  45000  45300  44300  44300     27400
1990-01-06  44800  45000  44500  44500     26380
1990-01-08  44500  44900  44000  44000     14790
...           ...    ...    ...    ...       ...
2022-03-17  71200  71800  70900  71200  17646315
2022-03-18  70600  70900  70200  70700  14410038
2022-03-21  70900  71000  69900  69900  11169002
2022-03-22  69900  70500  69900  70300   9402666
2022-03-23  70600  71200  70300  70500  12332242

[8373 rows x 5 columns]

In [89]:
# yahoo finance를 데이터 소스로 사용할 때는 종목코드 뒤에 코스피 종목인 경우 .KS, 코스닥 종목인 경우 .KQ를 붙여줘야 함
df_pdr = pdr.DataReader('005930.KS', 'yahoo', start=start_date, end=end_date)
df_pdr

High      Low     Open    Close      Volume     Adj Close
Date                                                                    
2000-01-04   6110.0   5660.0   6000.0   6110.0  74195000.0   4675.784180
2000-01-05   6060.0   5520.0   5800.0   5580.0  74680000.0   4270.191895
2000-01-06   5780.0   5580.0   5750.0   5620.0  54390000.0   4300.802246
2000-01-07   5670.0   5360.0   5560.0   5540.0  40305000.0   4239.580566
2000-01-10   5770.0   5580.0   5600.0   5770.0  46880000.0   4415.591797
...             ...      ...      ...      ...         ...           ...
2022-03-17  70400.0  70400.0  70400.0  70400.0         0.0  70400.000000
2022-03-18  70900.0  70200.0  70600.0  70700.0  14410038.0  70700.000000
2022-03-21  71000.0  69900.0  70900.0  69900.0  11169002.0  69900.000000
2022-03-22  70500.0  69900.0  69900.0  70300.0   9402666.0  70300.000000
2022-03-23  71200.0  70300.0  70600.0  70500.0  12315352.0  70500.000000

[5580 rows x 6 columns]

## 2.4. yfinance

+ yahoo finance를 데이터 소스로 사용하는 패키지이기 때문에 종목코드 뒤에 코스피 종목인 경우 .KS, 코스닥 종목인 경우 .KQ를 붙여줘야 함
+ 가장 다양한 정보 확인 가능
+ 2000년 데이터부터 조회됨
+ end에 설정한 일자의 전일자까지 조회하기 때문에 end에 +1일을 해서 넣어줘야함

In [101]:
ticker = yf.Ticker('005930.KS')

In [108]:
ticker.history(
               interval='1d',
               start=start_date,
               end='2022-03-24',
               actions=True,
               auto_adjust=True)

Open          High           Low         Close    Volume  \
Date                                                                           
2000-01-04   4591.605239   4675.784668   4331.414275   4675.784668  74195000   
2000-01-05   4438.550715   4637.520230   4224.275853   4270.191895  74680000   
2000-01-06   4400.287995   4423.246019   4270.192524   4300.803223  54390000   
2000-01-07   4254.886401   4339.065809   4101.832934   4239.581055  40305000   
2000-01-10   4285.496371   4415.591797   4270.191027   4415.591797  46880000   
...                  ...           ...           ...           ...       ...   
2022-03-17  70400.000000  70400.000000  70400.000000  70400.000000         0   
2022-03-18  70600.000000  70900.000000  70200.000000  70700.000000  14410038   
2022-03-21  70900.000000  71000.000000  69900.000000  69900.000000  11169002   
2022-03-22  69900.000000  70500.000000  69900.000000  70300.000000   9402666   
2022-03-23  70600.000000  71200.000000  70300.000000  70500.000000  12315352   

            Dividends  Stock Splits  
Date                                 
2000-01-04        0.0           0.0  
2000-01-05        0.0           0.0  
2000-01-06        0.0           0.0  
2000-01-07        0.0           0.0  
2000-01-10        0.0           0.0  
...               ...           ...  
2022-03-17        0.0           0.0  
2022-03-18        0.0           0.0  
2022-03-21        0.0           0.0  
2022-03-22        0.0           0.0  
2022-03-23        0.0           0.0  

[5580 rows x 7 columns]

In [107]:
# 분단위 데이터 조회. 조회할 수 있는 기간에 제한이 있음
# 마지막행의 데이터는 당일 종가로 조회되는 듯?
df_minute = ticker.history(
               interval='1m',
               start='2022-03-20',
               end='2022-03-22',
               actions=True,
               auto_adjust=True)
df_minute

Open     High      Low    Close  Volume  \
Datetime                                                                
2022-03-21 09:00:00+09:00  70900.0  71000.0  70700.0  70800.0       0   
2022-03-21 09:01:00+09:00  70800.0  70800.0  70500.0  70600.0  261699   
2022-03-21 09:02:00+09:00  70600.0  70600.0  70500.0  70600.0   53799   
2022-03-21 09:03:00+09:00  70500.0  70600.0  70500.0  70600.0   25665   
2022-03-21 09:04:00+09:00  70500.0  70600.0  70400.0  70400.0  281213   
...                            ...      ...      ...      ...     ...   
2022-03-21 14:56:00+09:00  70100.0  70100.0  70000.0  70000.0   45400   
2022-03-21 14:57:00+09:00  70000.0  70100.0  70000.0  70100.0   28528   
2022-03-21 14:58:00+09:00  70000.0  70100.0  69900.0  70000.0  374109   
2022-03-21 14:59:00+09:00  69900.0  70000.0  69900.0  69900.0   24122   
2022-03-23 15:00:00+09:00  70500.0  70500.0  70500.0  70500.0       0   

                           Dividends  Stock Splits  
Datetime                                            
2022-03-21 09:00:00+09:00          0             0  
2022-03-21 09:01:00+09:00          0             0  
2022-03-21 09:02:00+09:00          0             0  
2022-03-21 09:03:00+09:00          0             0  
2022-03-21 09:04:00+09:00          0             0  
...                              ...           ...  
2022-03-21 14:56:00+09:00          0             0  
2022-03-21 14:57:00+09:00          0             0  
2022-03-21 14:58:00+09:00          0             0  
2022-03-21 14:59:00+09:00          0             0  
2022-03-23 15:00:00+09:00          0             0  

[361 rows x 7 columns]

In [ ]:
# 재무정보도 조회가능
ticker.financials

2021-12-31         2020-12-31  \
Research Development                     22401726000000.0   21111490000000.0   
Effect Of Accounting Charges                         None               None   
Income Before Tax                        53351827000000.0   36345117000000.0   
Minority Interest                         8662234000000.0    8277685000000.0   
Net Income                               39243791000000.0   26090846000000.0   
Selling General Administrative           37084824000000.0   33127647000000.0   
Gross Profit                            113193457000000.0   92318692000000.0   
Ebit                                     51633856000000.0   35993876000000.0   
Operating Income                         51633856000000.0   35993876000000.0   
Other Operating Expenses                             None               None   
Interest Expense                          -431540000000.0    -583013000000.0   
Extraordinary Items                                  None               None   
Non Recurring                                        None               None   
Other Items                                          None               None   
Income Tax Expense                       13444377000000.0    9937285000000.0   
Total Revenue                           279604799000000.0  236806988000000.0   
Total Operating Expenses                227970943000000.0  200813112000000.0   
Cost Of Revenue                         166411342000000.0  144488296000000.0   
Total Other Income Expense Net            1717971000000.0     351241000000.0   
Discontinued Operations                              None               None   
Net Income From Continuing Ops           39907450000000.0   26407832000000.0   
Net Income Applicable To Common Shares   39243791000000.0   26090846000000.0   

                                               2019-12-31         2018-12-31  
Research Development                     19907236000000.0   18354080000000.0  
Effect Of Accounting Charges                         None               None  
Income Before Tax                        30432189000000.0   61159958000000.0  
Minority Interest                         7964949000000.0    7684184000000.0  
Net Income                               21505054000000.0   43890877000000.0  
Selling General Administrative           33444770000000.0   32688565000000.0  
Gross Profit                             83161332000000.0  111377004000000.0  
Ebit                                     27768509000000.0   58886669000000.0  
Operating Income                         27768509000000.0   58886669000000.0  
Other Operating Expenses                             None               None  
Interest Expense                          -686356000000.0    -674617000000.0  
Extraordinary Items                                  None               None  
Non Recurring                                        None               None  
Other Items                                          None               None  
Income Tax Expense                        8693324000000.0   16815101000000.0  
Total Revenue                           230400881000000.0  243771415000000.0  
Total Operating Expenses                202632372000000.0  184884746000000.0  
Cost Of Revenue                         147239549000000.0  132394411000000.0  
Total Other Income Expense Net            2663680000000.0    2273289000000.0  
Discontinued Operations                              None               None  
Net Income From Continuing Ops           21738865000000.0   44344857000000.0  
Net Income Applicable To Common Shares   21505054000000.0   43890877000000.0

# 3. 상장폐지 종목의 주가 조회

## 3.1. FinanceDataReader

In [109]:
krx_delisting[krx_delisting.Symbol=='001047']

Symbol   Name Market SecuGroup   Kind ListingDate DelistingDate Reason  \
6  001047  CJ2우B  KOSPI        주권  신형우선주  1999-01-29    2009-03-21          

  ArrantEnforceDate ArrantEndDate Industry  ParValue  ListingShares ToSymbol  \
6               NaT           NaT     서비스업    5000.0      1019041.0            

  ToName  
6

In [50]:
# KRX delisting stock data 상장폐지된 종목 가격 데이터 (상장일~상장폐지일)
df = fdr.DataReader('001047', exchange='KRX-DELISTING')
df

Date    Code       Name Market SecuGroup  Close  UpDown  Change  \
0    2009-03-19  001047      CJ2우B  KOSPI        주권  32650       0       0   
1    2009-03-18  001047      CJ2우B  KOSPI        주권  32650       0       0   
2    2009-03-17  001047      CJ2우B  KOSPI        주권  32650       0       0   
3    2009-03-16  001047      CJ2우B  KOSPI        주권  32650       0       0   
4    2009-03-13  001047      CJ2우B  KOSPI        주권  32650       0       0   
...         ...     ...        ...    ...       ...    ...     ...     ...   
2495 1999-02-04  001047  제일제당(2우B)  KOSPI        주권  14600       1    1000   
2496 1999-02-03  001047  제일제당(2우B)  KOSPI        주권  13600       2   -1300   
2497 1999-02-02  001047  제일제당(2우B)  KOSPI        주권  14900       1     100   
2498 1999-02-01  001047  제일제당(2우B)  KOSPI        주권  14800       2    -200   
2499 1999-01-29  001047  제일제당(2우B)  KOSPI        주권  15000       1    1400   

      ChangeRate   Open   High  Lower  Volume     Amount       MarCap  
0           0.00      0      0      0       0          0  33271688650  
1           0.00      0      0      0       0          0  33271688650  
2           0.00      0      0      0       0          0  33271688650  
3           0.00      0      0      0       0          0  33271688650  
4           0.00      0      0      0       0          0  33271688650  
...          ...    ...    ...    ...     ...        ...          ...  
2495        7.35  13600  14600  13600   17050  240588500  13882775000  
2496       -8.72  14800  15000  13600   26780  397083000  12931900000  
2497        0.68  14800  15000  14800   28480  424696000  14168037500  
2498       -1.33  14000  15100  13500   27720  406817000  14072950000  
2499       10.29  13600  15300  12000   47560  688146000  14263125000  

[2500 rows x 15 columns]

## 3.2. pykrx

In [52]:
stock.get_market_ohlcv_by_date(fromdate=start_date, 
                               todate=end_date,
                               ticker="001047")

시가     고가     저가     종가    거래량
날짜                                           
1999-01-29   6059   6816   5346   6682  47560
1999-02-01   6237   6727   6014   6593  27720
1999-02-02   6594   6683   6594   6637  28480
1999-02-03   6594   6683   6059   6058  26780
1999-02-04   6059   6505   6059   6504  17050
...           ...    ...    ...    ...    ...
2009-03-13  32650  32650  32650  32650      0
2009-03-16  32650  32650  32650  32650      0
2009-03-17  32650  32650  32650  32650      0
2009-03-18  32650  32650  32650  32650      0
2009-03-19      0      0      0  32650      0

[2500 rows x 5 columns]

## 3.3. pandas_datareader

In [53]:
df_pdr = pdr.DataReader('001047', 'naver', start=start_date, end=end_date)
df_pdr

Open   High    Low  Close Volume
Date                                         
1999-01-29   6059   6816   5346   6682  47560
1999-02-01   6237   6727   6014   6593  27720
1999-02-02   6594   6683   6594   6637  28480
1999-02-03   6594   6683   6059   6058  26780
1999-02-04   6059   6505   6059   6504  17050
...           ...    ...    ...    ...    ...
2009-03-13  32650  32650  32650  32650      0
2009-03-16  32650  32650  32650  32650      0
2009-03-17  32650  32650  32650  32650      0
2009-03-18  32650  32650  32650  32650      0
2009-03-19      0      0      0  32650      0

[2500 rows x 5 columns]

In [110]:
# yahoo finance에서는 상폐종목 조회가 안되는 듯
df_pdr = pdr.DataReader('001047.KS', 'yahoo', start=start_date, end=end_date)
df_pdr

KeyError: 'Date'

## 3.4. yfinance

In [111]:
ticker = yf.Ticker('001047.KS')

In [113]:
# yahoo finance에서는 상폐종목 조회가 안되는 듯
ticker.history(
               interval='1d',
               start=start_date,
               end='2022-03-24',
               actions=True,
               auto_adjust=True)

- 001047.KS: No data found for this date range, symbol may be delisted


Empty DataFrame
Columns: [Open, High, Low, Close, Adj Close, Volume]
Index: []

# 4. 패키지별 값 비교

In [114]:
start_date = '2000-01-01'
end_date = '2000-01-07'

In [115]:
df_fdr = fdr.DataReader('005930', start=start_date, end=end_date)
df_fdr

Open  High   Low  Close   Volume    Change
Date                                                  
2000-01-04  6000  6110  5660   6110  1483967  0.148496
2000-01-05  5800  6060  5520   5580  1493604 -0.086743
2000-01-06  5750  5780  5580   5620  1087810  0.007168
2000-01-07  5560  5670  5360   5540   806195 -0.014235

In [116]:
df_pykrx = stock.get_market_ohlcv_by_date(fromdate=start_date, 
                                          todate=end_date,
                                          ticker="005930")
df_pykrx

시가    고가    저가    종가      거래량
날짜                                         
2000-01-04  6000  6110  5660  6110  1483967
2000-01-05  5800  6060  5520  5580  1493604
2000-01-06  5750  5780  5580  5620  1087810
2000-01-07  5560  5670  5360  5540   806195

In [117]:
df_pdr = pdr.DataReader('005930', 'naver', start=start_date, end=end_date)
df_pdr

Open  High   Low Close   Volume
Date                                       
2000-01-04  6000  6110  5660  6110  1483967
2000-01-05  5800  6060  5520  5580  1493604
2000-01-06  5750  5780  5580  5620  1087810
2000-01-07  5560  5670  5360  5540   806195

In [118]:
df_pdr = pdr.DataReader('005930.KS', 'yahoo', start=start_date, end=end_date)
df_pdr

High   Low  Open  Close    Volume    Adj Close
Date                                                      
2000-01-04  6110  5660  6000   6110  74195000  4675.783203
2000-01-05  6060  5520  5800   5580  74680000  4270.191895
2000-01-06  5780  5580  5750   5620  54390000  4300.803223
2000-01-07  5670  5360  5560   5540  40305000  4239.580566

In [119]:
ticker = yf.Ticker('005930.KS')

In [122]:
# yahoo finance에서는 상폐종목 조회가 안되는 듯
ticker.history(
               interval='1d',
               start=start_date,
               end='2000-01-08',
               actions=True,
               auto_adjust=False)

Open    High     Low   Close    Adj Close    Volume  Dividends  \
Date                                                                           
2000-01-04  6000.0  6110.0  5660.0  6110.0  4675.782715  74195000          0   
2000-01-05  5800.0  6060.0  5520.0  5580.0  4270.191895  74680000          0   
2000-01-06  5750.0  5780.0  5580.0  5620.0  4300.801758  54390000          0   
2000-01-07  5560.0  5670.0  5360.0  5540.0  4239.581055  40305000          0   

            Stock Splits  
Date                      
2000-01-04             0  
2000-01-05             0  
2000-01-06             0  
2000-01-07             0

+ 거래량이 yahoo finance를 데이터소스로 사용하는 API들만 다름.
+ 증권플러스(두나무에서 개발)에서 조회해보면 yahoo finance 거래량에 더 가깝긴한데 일치하지는 않음
    + 2000-01-04: 72,037,378
    + 2000-01-05: 72,504,854
    + 2000-01-06: 52,806,310
    + 2000-01-07: 39,135,679

+ 미래에셋증권에서 조회
    + 2000-01-04: 74,084,850
    + 2000-01-05: 74,680,200
    + 2000-01-06: 54,390,500
    + 2000-01-07: 40,309,750

+ NH투자증권에서 조회
    + 2000-01-04: 74,198,350
    + 2000-01-05: 74,680,200
    + 2000-01-06: 54,390,500
    + 2000-01-07: 40,309,750
    
+ 키움증권에서 조회
    + 2000-01-04: 74,198,350
    + 2000-01-05: 74,680,200
    + 2000-01-06: 54,390,500
    + 2000-01-07: 40,309,750

+ 가격 데이터는 증권플러스랑은 다르고 미래에셋, NH투자증권, 키움증권과 같음
+ krx에서 확인해보니 NH랑 키움 값이 맞음 - 1483967*50=74198350
+ 네이버금융은 거래량에 액면분할 고려 안되어 있음